src/Security/Authentication/OAuth/src/Events/OAuthEvents.cs

```diff
@@ -25,6 +29,41 @@ public class OAuthEvents : RemoteAuthenticationEvents
             return Task.CompletedTask;
         };
 
+        /// <summary>
+        /// Gets or sets the delegate that is invoked when the ExchangeCode method is invoked.
+        /// </summary>
+        public Func<OAuthExchangeCodeContext, Task<OAuthTokenResponse>> OnExchangeCode { get; set; } = async context =>
```

Comment:
This is more code than is appropriate for events. Deriving from OAuthHandler is more appropriate once you need to replace this much.

Is there a more specific event that would still be useful? Like adding an event on line 45 that passed the parameters dictionary so it could be augmented? The default implementation would no-op. Would that be enough for your PKCE example?